packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+    });
+  }
+
+  /**
+   * Constructs the requested caveats for the permission of the specified
+   * subject origin and target name.
+   *
+   * Throws an error if validation fails.
+   *
+   * @see {@link PermissionController.constructCaveat} for details.
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveats - The requested caveats to construct.
+   * @returns The constructed caveats.
+   */
+  private constructCaveats(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveats?: unknown[] | null,
+  ): NonEmptyArray<ExtractCaveats<ControllerCaveatSpecification>> | undefined {
+    const caveatArray = requestedCaveats?.map((requestedCaveat) =>
+      this.constructCaveat(origin, target, requestedCaveat),
+    );
+    return caveatArray && isNonEmptyArray(caveatArray)
+      ? caveatArray
+      : undefined;
+  }
+
+  /**
+   * Constructs a caveat for the permission of the specified subject origin and
+   * target name, per the requested caveat argument. This methods validates
+   * everything about the requested caveat, except that its `value` property
+   * is JSON-compatible. It also ensures that a caveat specification exist for
+   * the requested type, and calls the specification validator, if it exists, on
+   * the constructed caveat.
+   *
+   * Throws an error if validation fails.
+   *
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveat - The requested caveat to construct.
+   * @returns The constructed caveat.
+   */
+  private constructCaveat(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveat: unknown,
+  ): ExtractCaveats<ControllerCaveatSpecification> {
+    if (!isPlainObject(requestedCaveat)) {
+      throw new InvalidCaveatError(requestedCaveat, origin, target);
+    }
+
+    if (Object.keys(requestedCaveat).length !== 2) {
+      throw new InvalidCaveatFieldsError(requestedCaveat, origin, target);
+    }
+
+    const { type, value } = requestedCaveat;
+    if (typeof type !== 'string') {
+      throw new InvalidCaveatTypeError(requestedCaveat, origin, target);
+    }
+
+    if (!hasProperty(requestedCaveat, 'value')) {
+      throw new CaveatMissingValueError(requestedCaveat, origin, target);
+    }
+
+    // Assert that the caveat value is valid JSON
+    if (!deepEqual(value, JSON.parse(JSON.stringify(value)))) {
+      throw new CaveatInvalidJsonError(requestedCaveat, origin, target);
+    }
+
+    const caveat = _constructCaveat(type, (value as Json) ?? null);
+    this.validateCaveat(caveat, origin, target);
+    return caveat as ExtractCaveats<ControllerCaveatSpecification>;
```

Comment:
Nit: If we update `validateCaveat` to return `true` for valid caveats and add a type predicate, we can get rid of this `as` as well.

It's a bit awkward because there's no `false` case though :thinking: We'd still have to do something like `return undefined as never` to cover the `false` case. Not sure if there's a nicer way to do this or not.

Reply:
I use this validator pattern of "if the function returns the thing is valid, otherwise it will throw" in a handful of places. I'm also not sure if there's a nicer / better way of doing it. The problem with doing an `if`/`else` with a type guard is that a type guard doesn't say _why_ the value is invalid if it returns false.

Using type guards, if we want to have specific errors, we have to effectively replicate the logic of the type guard so we can throw a specific error. That's just a mess, and while we don't like casts, they seem warranted for this kind of validator.

Reply:
Fair enough. It seems OK in cases like this where the cast is right next to the validator, where it should be clear to the reader why the cast is warranted.

I wish type predicates could be applied to functions of this style, where they throw if invalid.